# Data Science I
### Klausur I im Sommersemester 2024

## Allgemeine Informationen

* Sie haben eine Woche Zeit, um die Klausur zu bearbeiten.

* Sie können alle Quellen verwenden, müssen sie jedoch korrekt benennen. Wenn Sie ChatGPT oder eine ähnliche Software verwenden, müssen Sie dies kenntlich machen und den verwendeten Prompt angeben.

* Sie sollten die folgenden Pakete verwenden: `numpy, pandas, scipy, geopy, scikit-learn/sklearn, matplotlib, seborn, openPyxl` und Pythons Standardlibraries. Diese sind ausreichend, um die Klausur zu lösen. Falls Sie andere Pakete verwenden, rechtfertigen Sie deren Verwendung.

* Der Code muss ausreichend kommentiert und verständlich sein. Schreiben Sie Funktionen beim Wiederverwenden von Code. Befolgen Sie im Allgemeinen die Richtlinien aus der Vorlesung. Punkte können aufgrund eines schlecht strukturierten oder unverständlichen Codes abgezogen werden.

* **Begründen Sie Entscheidungen** zur Auswahl von Plots, Hypothesentest usw. und **interpretieren Sie** Ihre Ergebnisse.

* Sie dürfen in keiner Form Hilfe oder Rat von Dritten in Anspruch nehmen.

* Bitte laden Sie Ihre vollständige Lösung der Klausur als `.zip`-Datei mit dem Dateinamen `vorname_matrikelnummer.zip` bis 8. August 2024 um 12:00 Uhr auf StudIP in den Ordner `Submission - Exam 1` hoch.

* Fügen Sie der `.zip` Datei auch die unterschriebene Eigenständigkeitserklärung hinzu.

* Wenn Sie Fragen haben, kontaktieren Sie uns bitte rechtzeitig über Rocketchat.

In [4]:
# IMPORT LIBRARIES
import numpy as np, pandas as pd, scipy, geopy, sklearn, matplotlib, seaborn, openpyxl, warnings, os 

warnings.filterwarnings("ignore", category=UserWarning, module='openpyxl')

# Debug Level
# 1 - Loading Data
# 2 - Change Loaded Data
debug = 1

## Aufgaben und Punkte:

<table>
  <thead>
    <tr>
      <th colspan="3">Aufgabe 1 - Data Preprocessing</th>
      <th colspan="2">Aufgabe 2 - Plotting</th>
      <th colspan="2">Aufgabe 3 - Statistics</th>
      <th colspan="2">Aufgabe 4 - Machine Learning </th>
    </tr>
    <tr>
      <th>Aufgabe 1.1</th>
      <th>Aufgabe 1.2</th>
      <th>Aufgabe 1.3</th>
      <th>Aufgabe 2.1</th>
      <th>Aufgabe 2.2</th>
      <th>Aufgabe 3.1</th>
      <th>Aufgabe 3.2</th>
      <th>Aufgabe 4.1</th>
      <th>Aufgabe 4.2</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>10 Punkte </td>
      <td>12 Punkte </td>
      <td>2 Punkte </td>
      <td>11 Punkte</td>
      <td>27 Punkte </td>
      <td>13 Punkte </td>
      <td>5 Punkte </td>
      <td>10 Punkte </td>
      <td>10 Punkte </td>
    </tr>
    <!-- Add more rows as needed -->
  </tbody>
</table>


_____
## Aufgabe 0: Setup

Der Klausurordner enthält ein `Dockerfile`, in dem alle relevanten Pakete definiert sind. Das `Dockerfile` baut auf dem Jupyter Server Image auf. Verwenden Sie dieses Dockerfile, um zuerst ein Docker Image zu erstellen und dann einen Docker Container von diesem Image zu starten. Benutzen Sie anschließend die Jupyter Server Instanz, um an der Klausur zu arbeiten. Wir empfehlen dringend, die Docker-Umgebung zu verwenden, um Versionskonflikte zwischen den verschiedenen Paketen zu vermeiden. Code, der in dieser Umgebung nicht ausführbar ist, wird als **nicht funktional** bewertet.

____
## Aufgabe 1: Data Preprocessing (24 Punkte)

### Datenbeschreibung

Im Ordner `data` finden Sie die monatlichen Parkdaten der Stadt Göttingen für das Jahr 2023 (Feb.-Dez.). Die Parkschein-Verkäufe an den stationären Parkscheinautomaten befinden sich in den Dateien, deren Namen mit `Cale` beginnt, und die mit der Parkster-App gekauften Parkscheine befinden sich in den Dateien, deren Namen mit `Parkster` beginnen.<br>
Die Datei `parkzone_latlong.csv` enthält weitere geografische Informationen zu den Parkzonen und die Datei `psa_latlong.csv` enthält geografische Informationen über die Parkscheinautomaten innerhalb der Parkzone.

Die bereitgestellten Parkdaten sind echte Rohdaten und stammen direkt von der Stadt Göttingen. Wir haben lediglich die geografischen Informationen hinzugefügt.

*Bitte beachten Sie:*
- *Obwohl wir nur Daten von Februar bis Dezember haben, bezeichnen wir diese im Folgenden als jährlich.*
- *Aufgrund der Größe der Daten sollten Sie Ihren Arbeitsspeicher effizient verwenden. Vermeiden Sie daher die Speicherung mehrerer Kopien desselben DataFrames.*

#### Aufagbe 1.1 - Laden der Daten (10 Punkte)
Laden Sie die Dateien für die Parkscheinautomaten (`Cale-*`) und für die App (`Parkster-*`) und fügen Sie diese **jeweils** zu einem Dataframe zusammen, der die jährlichen Verkäufe für Parkscheinautomaten und App beinhaltet. <br>
Laden Sie auch die weiteren Informationen zu den Parkscheinautomaten (`psa_latlong.csv`) und Parkzonen (` parkzones_latlong.csv`).

Sie werden die Werte `0` und `999` in der Spalte `Automaten -ID` für die Daten der Parkscheinautomaten finden. <br> Ändern Sie die `0`en in `1`en und löschen Sie alle Einträge mit `999`.
Überprüfen Sie auch auf Zeilen-Duplikate und löschen Sie diese gegebenenfalls. 

In [5]:
# Function to load multiple datafiles as one dataframe
def load_data_files(filtername, skippingrows=0): # skippingrows are rows to skip
    print(f"{filtername} datafiles loading...")
    data_files = list(filter(lambda x: x.startswith(filtername), os.listdir('data'))) # file searching
    data_files_df = pd.concat([pd.read_excel(f"data/{file}", skiprows=skippingrows) for file in data_files]) # combine files in dataframe
    print(f"{filtername} datafiles finish\n===============")
    if debug <= 1: print(data_files_df)
    return data_files, data_files_df # return tuple of the files and the dataframe

# Function to load file
def loadcsv(filename):
    print(f"{filename} loading")
    file_df = pd.read_csv(f'data/{filename}')
    print(f"{filename} finish\n===============")
    if debug <= 1: print(file_df)
    return file_df

In [6]:
# data_loading "Cale-" skipping row 0 and 1 because of different formatting
parking_meters, parking_meters_df = load_data_files("Cale-", 2)

Cale- datafiles loading...
Cale- datafiles finish
      Automat - Automaten ID Zahleinheit - Name     Knoten  \
0                      PA320              Münze  Göttingen   
1                       PA03              Münze  Göttingen   
2                      PA354              Münze  Göttingen   
3                      PA332              Münze  Göttingen   
4                      PA334              Münze  Göttingen   
...                      ...                ...        ...   
68207                   PA85              Münze  Göttingen   
68208                   PA85              Münze  Göttingen   
68209                   PA85              Münze  Göttingen   
68210                   PA85              Münze  Göttingen   
68211                   PA84              Münze  Göttingen   

           Kaufdatum Lokal  Betrag Artikelname  Artikel ID Tarifpaket - Name  \
0      01.11.2023 00:28:56     1.9  Parkticket           0      Normalticket   
1      01.11.2023 00:44:21     4.0  Parkticke

In [7]:
# data_loading "Parkster-"
parkzones, parkzones_df = load_data_files("Parkster-")

Parkster- datafiles loading...
Parkster- datafiles finish
       Parkzone             Erstellt                Start  \
0         37008  2023-01-31 13:07:44  2023-01-31 13:07:44   
1         37005  2023-01-31 15:40:23  2023-01-31 15:40:23   
2         37005  2023-01-31 15:48:59  2023-01-31 15:48:59   
3         37204  2023-01-31 15:58:47  2023-01-31 15:58:47   
4         37105  2023-01-31 16:08:01  2023-01-31 16:08:00   
...         ...                  ...                  ...   
40050     37108  2023-12-31 16:45:25  2023-12-31 16:45:25   
40051     37001  2023-12-31 17:09:16  2023-12-31 17:09:16   
40052     37105  2023-12-31 17:25:59  2023-12-31 17:25:59   
40053     37001  2023-12-31 17:52:09  2023-12-31 17:52:09   
40054     37108  2023-12-31 20:40:11  2023-12-31 20:40:11   

                     Stopp Parkgebühren inkl. MwSt. in EUR     Status  \
0      2023-02-01 14:00:00                            8,31     normal   
1      2023-02-01 11:45:19                            4,96     

In [8]:
parkzones_latlong_df = loadcsv("parkzones_latlong.csv")

parkzones_latlong.csv loading
parkzones_latlong.csv finish
     latitude  longitude  Zonencode  \
0   51.534248   9.936501      37001   
1   51.531270   9.936706      37005   
2   51.541762   9.942282      37008   
3   51.541249   9.944294      37009   
4   51.553603   9.942338      37010   
5   51.529542   9.939956      37101   
6   51.533104   9.926817      37102   
7   51.534678   9.942051      37103   
8   51.536634   9.932465      37104   
9   51.529205   9.933063      37105   
10  51.530127   9.937980      37106   
11  51.531365   9.927962      37107   
12  51.536733   9.928201      37108   
13  51.537675   9.930298      37109   
14  51.533859   9.943545      37201   
15  51.528115   9.940574      37202   
16  51.528164   9.940557      37203   
17  51.540630   9.935538      37204   
18  51.537585   9.940781      37205   
19  51.541593   9.937814      37206   
20  51.528780   9.931563      37207   
21  51.532833   9.935181      37208   
22  51.528850   9.936085      37209   

    

In [9]:
psa_latlong_df = loadcsv("psa_latlong.csv")

psa_latlong.csv loading
psa_latlong.csv finish
       PSA                 location   latitude  longitude   zone
0        1           Am Geismar Tor  51.529542   9.939956  37101
1    1 - 2           Am Geismar Tor  51.529542   9.939956  37101
2        2             Bürgerstraße  51.533104   9.926817  37102
3        3                Papendiek  51.533827   9.931693  37001
4        4       Bonifatiusschule -  51.528850   9.936085  37209
..     ...                      ...        ...        ...    ...
199    358          Kreuzbergring 2  51.542319   9.932972  37008
200    359          Kreuzbergring 3  51.542898   9.933473  37008
201    360  Von - Siebold- Straße 1  51.545994   9.943523  37008
202    361  Von - Siebold- Straße 2  51.546472   9.943674  37008
203    362  Von - Siebold- Straße 3  51.545665   9.942838  37008

[204 rows x 5 columns]


In [10]:

# ID Sortierung in Ints um bessere Übersicht zu haben.
parking_meters_df['Automat - Automaten ID'] = parking_meters_df['Automat - Automaten ID'].str.replace("PA", "").astype(int)
if debug <= 2: print(parking_meters_df)

       Automat - Automaten ID Zahleinheit - Name     Knoten  \
0                         320              Münze  Göttingen   
1                           3              Münze  Göttingen   
2                         354              Münze  Göttingen   
3                         332              Münze  Göttingen   
4                         334              Münze  Göttingen   
...                       ...                ...        ...   
68207                      85              Münze  Göttingen   
68208                      85              Münze  Göttingen   
68209                      85              Münze  Göttingen   
68210                      85              Münze  Göttingen   
68211                      84              Münze  Göttingen   

           Kaufdatum Lokal  Betrag Artikelname  Artikel ID Tarifpaket - Name  \
0      01.11.2023 00:28:56     1.9  Parkticket           0      Normalticket   
1      01.11.2023 00:44:21     4.0  Parkticket           0      Normalticket   
2  

In [11]:
# Convert MwSt to float
parkzones_df['Parkgebühren inkl. MwSt. in EUR'] = parkzones_df['Parkgebühren inkl. MwSt. in EUR'].str.replace(',', '.').astype(float) 
if debug <= 2: print(parkzones_df)

       Parkzone             Erstellt                Start  \
0         37008  2023-01-31 13:07:44  2023-01-31 13:07:44   
1         37005  2023-01-31 15:40:23  2023-01-31 15:40:23   
2         37005  2023-01-31 15:48:59  2023-01-31 15:48:59   
3         37204  2023-01-31 15:58:47  2023-01-31 15:58:47   
4         37105  2023-01-31 16:08:01  2023-01-31 16:08:00   
...         ...                  ...                  ...   
40050     37108  2023-12-31 16:45:25  2023-12-31 16:45:25   
40051     37001  2023-12-31 17:09:16  2023-12-31 17:09:16   
40052     37105  2023-12-31 17:25:59  2023-12-31 17:25:59   
40053     37001  2023-12-31 17:52:09  2023-12-31 17:52:09   
40054     37108  2023-12-31 20:40:11  2023-12-31 20:40:11   

                     Stopp  Parkgebühren inkl. MwSt. in EUR     Status  \
0      2023-02-01 14:00:00                             8.31     normal   
1      2023-02-01 11:45:19                             4.96     normal   
2      2023-02-01 11:34:59                   

In [12]:
# 0s in 1s
parking_meters_df['Automat - Automaten ID'] = parking_meters_df['Automat - Automaten ID'].replace(0, 1)

# Drop Entry 999
parking_meters_df = parking_meters_df[parking_meters_df['Automat - Automaten ID'] != 999]

# Drop Duplicates
parking_meters_df.drop_duplicates()
parkzones_df.drop_duplicates()



Parkzone             Erstellt                Start  \
0         37008  2023-01-31 13:07:44  2023-01-31 13:07:44   
1         37005  2023-01-31 15:40:23  2023-01-31 15:40:23   
2         37005  2023-01-31 15:48:59  2023-01-31 15:48:59   
3         37204  2023-01-31 15:58:47  2023-01-31 15:58:47   
4         37105  2023-01-31 16:08:01  2023-01-31 16:08:00   
...         ...                  ...                  ...   
40050     37108  2023-12-31 16:45:25  2023-12-31 16:45:25   
40051     37001  2023-12-31 17:09:16  2023-12-31 17:09:16   
40052     37105  2023-12-31 17:25:59  2023-12-31 17:25:59   
40053     37001  2023-12-31 17:52:09  2023-12-31 17:52:09   
40054     37108  2023-12-31 20:40:11  2023-12-31 20:40:11   

                     Stopp  Parkgebühren inkl. MwSt. in EUR     Status  \
0      2023-02-01 14:00:00                             8.31     normal   
1      2023-02-01 11:45:19                             4.96     normal   
2      2023-02-01 11:34:59                             4.74     normal   
3      2023-02-01 15:44:00                             7.53     normal   
4      2023-02-01 01:38:00                             4.30     normal   
...                    ...                              ...        ...   
40050  2023-12-31 17:43:25                              NaN     normal   
40051  2023-12-31 19:39:16                              NaN     normal   
40052  2023-12-31 19:45:35                              NaN     normal   
40053  2023-12-31 17:52:28                              NaN  storniert   
40054  2023-12-31 23:10:11                              NaN     normal   

      Parkscheinart  Zonencode Eigentümercode  
0          Kurzzeit      37008                 
1          Kurzzeit      37005                 
2          Kurzzeit      37005                 
3          Kurzzeit      37204                 
4          Kurzzeit      37105                 
...             ...        ...            ...  
40050      Kurzzeit      37108                 
40051      Kurzzeit      37001                 
40052      Kurzzeit      37105                 
40053      Kurzzeit      37001                 
40054      Kurzzeit      37108                 

[381278 rows x 9 columns]

#### Aufgabe 1.2 - Zusammenführen und Formatieren (12 Punkte)
Erstellen Sie einen DataFrame für beide Verkaufsarten, indem Sie die beiden zuvor erstellen DataFrames zusammenführen. Nutzen Sie dazu die Parkzonen Informationen *(in `parkzones_latlong.csv`)* und die Parkscheinautomatennummer *(in` pa_latlong.csv`)*. Stellen Sie sicher, dass sich in Ihrem DataFrame die geografischen Informationen für Parkscheinautomaten und Parkzonen befinden.
Verwenden Sie die Spalten `Kaufdatum Lokal` und `Start` für das Kaufdatum, codieren Sie die Spalte als `datetime`-Objekt und verwenden Sie sie als Indexspalte. Stellen Sie außerdem sicher, dass die anderen Spalten ein angemessenes Datenformat haben.

*Hinweis: Es ist zu erwarten, dass `Nan`-Werte für einige Spalten in den Zeilen zu Appkäufen auftauchen.*

In [13]:
# DataStructure Creation
park_data = {
    'time': 'datetime64[ns]',
    'machine_ID': 'Int64',
    'fee': 'float64',
    'category': 'object',
    'street': 'object',
    'latitude_machine': 'float64',
    'longitude_machine': 'float64',
    'zone': 'int64',
    'latitude_zone': 'float64',
    'longitude_zone': 'float64'}

In [14]:
# Um das mergen einfacher zu machen
parking_meters_df['Automat - Automaten ID'] = parking_meters_df['Automat - Automaten ID'].astype(str)
psa_latlong_df['PSA'] = psa_latlong_df['PSA'].astype(str)
parkzones_df['Zonencode'] = parkzones_df['Parkzone'].astype(str)

In [15]:
merged_psa_df = pd.merge(parking_meters_df, psa_latlong_df, how='left', left_on='Automat - Automaten ID', right_on='PSA')

In [16]:
merged_psa_df['Kaufdatum Lokal'] = pd.to_datetime(merged_psa_df['Kaufdatum Lokal'], format='%d.%m.%Y %H:%M:%S')

In [21]:
# Parkscheinautomaten-Daten mit geografischen Informationen zusammenführen
merged_psa_df = pd.merge(parking_meters_df, psa_latlong_df, how='left', left_on='Automat - Automaten ID', right_on='PSA')


# Parkzonen-Daten mit geografischen Informationen zusammenführen
merged_parkzones_df = pd.merge(parkzones_df, parkzones_latlong_df, how='left', left_on='Parkzone', right_on='Zonencode')

In [22]:
# Spalten in datetime-Objekte konvertieren
merged_psa_df['Kaufdatum Lokal'] = pd.to_datetime(merged_psa_df['Kaufdatum Lokal'], format='%d-%m-%Y %H:%M:%S')
merged_parkzones_df['Start'] = pd.to_datetime(merged_parkzones_df['Start'], format='%Y-%m-%d %H:%:M:%S')

ValueError: ':' is a bad directive in format '%Y-%m-%d %H:%:M:%S'

#### Aufgabe 1.3 - DataFrame Check (2 Punkte)
Der bereinigte und vollständige DataFrame für die folgenden Aufgaben sollte der Datei `data/clean_dataframe.csv` entsprechen, der wie folgt eingelesen werden kann:

In [ ]:
pd.read_csv('data/clean_dataframe.csv', parse_dates=['time'], index_col='time', dtype={'machine_ID': 'Int64', 
                                                                           'fee': 'float64', 
                                                                           'category': 'object', 
                                                                           'street': 'object', 
                                                                           'latitude_machine': 'float64', 
                                                                           'longitude_machine': 'float64', 
                                                                           'zone': 'int64', 
                                                                           'latitude_zone': 'float64', 
                                                                           'longitude_zone': 'float64'}).sort_index()

machine_ID   fee category street  latitude_machine  \
time                                                                      
2023-01-31 13:07:44        <NA>  8.31      app    NaN               NaN   
2023-01-31 15:40:23        <NA>  4.96      app    NaN               NaN   
2023-01-31 15:48:59        <NA>  4.74      app    NaN               NaN   
2023-01-31 15:58:47        <NA>  7.53      app    NaN               NaN   
2023-01-31 16:08:00        <NA>  4.30      app    NaN               NaN   
...                         ...   ...      ...    ...               ...   
2023-12-31 16:45:25        <NA>  0.00      app    NaN               NaN   
2023-12-31 17:09:16        <NA>  0.00      app    NaN               NaN   
2023-12-31 17:25:59        <NA>  0.00      app    NaN               NaN   
2023-12-31 17:52:09        <NA>  0.00      app    NaN               NaN   
2023-12-31 20:40:11        <NA>  0.00      app    NaN               NaN   

                     longitude_machine   zone  latitude_zone  longitude_zone  
time                                                                          
2023-01-31 13:07:44                NaN  37008      51.541762        9.942282  
2023-01-31 15:40:23                NaN  37005      51.531270        9.936706  
2023-01-31 15:48:59                NaN  37005      51.531270        9.936706  
2023-01-31 15:58:47                NaN  37204      51.540630        9.935538  
2023-01-31 16:08:00                NaN  37105      51.529205        9.933063  
...                                ...    ...            ...             ...  
2023-12-31 16:45:25                NaN  37108      51.536733        9.928201  
2023-12-31 17:09:16                NaN  37001      51.534248        9.936501  
2023-12-31 17:25:59                NaN  37105      51.529205        9.933063  
2023-12-31 17:52:09                NaN  37001      51.534248        9.936501  
2023-12-31 20:40:11                NaN  37108      51.536733        9.928201  

[1089846 rows x 9 columns]

Stellen Sie sicher, dass Ihr DataFrame mit `clean_dataframe.csv` übereinstimmt. Verwenden Sie dazu die Funktion [`pandas.DataFrame.equals`](https://pandas.pydata.org/docs/reference/api/pandas.dataframe.equals.html).

Sollte `pandas.DataFrame.equals` nach Ihren Anpassungen nicht `True` zurückgeben, arbeiten Sie bitte mit `clean_dataframe.csv` weiter und geben Sie dies in einer Markdown-Zelle an. In diesem Fall erhalten Sie keine Punkte für die Teilaufgabe 1.3.

_____
## Aufgabe 2: Plotting (38 Punkte)



### Aufgabe 2.1 - Analyse der Parkscheinautomaten (11 Punkte)
Die Stadt Göttingen möchte einen Überblick über die Umsätze der einzelnen **Parkscheinautomaten** erhalten und stellt Sie für eine anfängliche explorative Analyse des Verkaufsvolumens und der geografischen Anordnung der Automaten ein.

#### 2.1.1 (6 Punkte)
Finden Sie die fünf umsatzstärksten Parkscheinautomaten im Jahr 2023 und visualisieren Sie den **wöchentlichen** Umsatz im Laufe des Jahres.

#### 2.1.2 (5 Punkte)
Der Standort der Parkscheinautomaten könnte auch einen Einfluss auf deren Umsatz haben.

Machen Sie sich mit der Funktion `plot_map` aus der Bibliothek `dsplotter` vertraut. Verwenden Sie die Funktion, um den jährlichen Umsatz für jeden Automaten auf einer Karte zu visualisieren. Machen Sie die Farbe **und** den Radius der Standortmarkierung abhängig vom jährlichen Umsatz. Was haben Automaten mit einem hohen jährlichen Umsatz gemeinsam?

### Aufgabe 2.2 - Analyse der Automaten- und Appnutzung pro Parkzone (27 Punkte)
Im Rahmen der Digitalisierungsinitiative der Stadt wurde die Parkster-App vor einigen Jahren als Alternative zu Parkscheinautomaten eingeführt.
Bisher wurden nur die Parkscheinautomaten in die Analyse eingeschlossen und daher einen Großteil des Ticketverkaufs, der über die App stattfand, nicht beachtet.

Die Stadt möchte für 2023 eine erste visuelle Analyse der Akzeptanz der App in den einzelnen Parkzonen durchführen und anschließend den gesamten Umsatz analysieren.

*Hinweis: Beachten Sie, dass wir die Umsätze aus der Appnutzung nur einschließen können, indem wir die Gesamtauswertung auf Ebene der Parkzonen durchführen.*

#### 2.2.1 (6 Punkte)
Bevorzugen Parkende die App- oder die Parkscheinautomatennutzung? 

Verwenden Sie einen geeigneten Plot, um die durchschnittliche Automaten- bzw. Appnutzungsrate pro **Parkzone** für das gesamte Jahr 2023 zu visualisieren. Was können Sie dem Plot entnehmen?

#### 2.2.2 (9 Punkte)
Wie stark werden die einzelnen Parkzonen genutzt? 

Visualisieren Sie die Gesamtzahl der Verkäufe und die Automaten- bzw. Appnutzungrate für jede Parkzone im Jahr 2023 **in einem Diagramm**. Verwenden Sie für die y-Achse eine `log`-Skalierung. Was können Sie dem Plot entnehmen?

#### 2.2.3 (7 Punkte)
Der vorherige Plot gibt uns eine Vorstellung von der Gesamtzahl der Parktickets *pro Parkzone*. Diese korreliert sehr wahrscheinlich stark mit der Anzahl der Parkplätze pro Zone. Um Parkzonen mit einer unterschiedlichen Anzahl an Parkplätzen zu vergleichen, sollten wir diese mithilfe der Anzahl der verfügbaren Parkplätze je Zone relativieren. Auf diese Weise können wir herausfinden, welche Zonen, relativ zu ihrer Größe, am häufigsten verwendet werden. Da wir nicht die Anzahl der Parkplätze für jede Zone zur Verfügung haben, können wir nur die Anzahl der Parkscheinautomaten als grobe Annäherung verwenden. 

Verwenden Sie die Informationen aus `psa_latlong.csv` und reproduzieren Sie den vorherigen mit Plot der Gesamtzahl der Verkäufe pro Automat für jede Parkzone. Verwenden Sie für die y-Achse eine `log`-Skalierung. Welche Parkzone wird am meisten genutzt?

#### 2.2.4 (5 Punkte)
Bisher haben wir die geografischen Informationen der Parkzonen nicht mit einbezogen. 

Verwenden Sie erneut die Funktion `plot_map`, um den Standort aller Parkzonen, ihre durchschnittlichen Tickets pro Automat und die Automaten- bzw. Appnutzungsrate zu visualieren. Färben Sie den Kartenmarker mithilfe der Automaten- bzw. Appnutzungsrate und legen Sie den Radius mit den durschnittlich verkauften Tickets pro Automat fest. Was können Sie der Darstellung entnehmen?

___
## Aufgabe 3: Statistics (18 Punkte)


#### Aufgabe 3.1 - t-Test (13 Punkte)
Zusätzlich zu der visuellen Analyse möchte die Stadt nun auch eine statistische Untersuchung der Verwendung von Parkscheinautomaten und Apps durchführen.

Bestimmen Sie dazu zunächst die Automaten- bzw. Appnutzungsrate pro Parkzone für jeden Kalendertag. Führen Sie anschließend für **jede Parkzone** einen t-Test durch, der testet, ob Parkende es vorziehen die App in der jeweiligen Zone zu verwenden. Schreiben Sie das entsprechende Hypothesenpaar auf, führen Sie den Test durch und interpretieren Sie Ihre Testergebnisse. Verwenden Sie für Ihre Testentscheidung ein Signifikanzniveau von 0.05. Welche grundlegende Annahme von statistischen Tests könnte bei diesem Vorgehen verletzt werden?

#### Aufgabe 3.2 - Statistisches Verständnis (5 Punkte)
Angenommen, für die Zone `37106` beträgt die durchschnittliche Automaten- bzw. Appnutzungsrate `0.5`.
Die Stadt sendet Ihnen die Daten für 2024. 

An wie vielen Tagen können Sie erwarten, dass die Appnutzung signifikant höher ist, wenn Sie weiterhin von einem Signifikanzniveau von `0.05` ausgehen? Erklären Sie, warum dies der Fall ist. Nehmen Sie an, dass sich das Verhalten der Parkenden im Vergleich zu 2023 nicht geändert hat.

___
## Aufgabe 4: Machine Learning (20 Punkte)

Nutzen Sie ein K-Nearest-Neighbors (KNN) Modell, um die Automaten- bzw. Appnutzungsrate mithilfe des Standortes (`latitude`, `longitude`) und der Parkgebühr (`fee`) vorherzusagen. Verwenden Sie ausschließlich Datenreihen mit Parkgebühren zwischen 2 Euro und 7 Euro.

#### Aufgabe 4.1 - Modell-Training and Hyperparameter-Suche (10 Punkte)
Bereiten Sie die Daten sinnvoll auf, führen Sie eine Hyperparameter-Suche nach optimalem K-Wert aus, visualisieren Sie die Ergebnisse der Hyperparameter-Suche und verwenden Sie schließlich Ihren optimalen K-Wert, um das Modell zu trainieren.

*Hinweis: Verwenden Sie 30% aller Daten zur Bestimmung des optimalen K-Wertes, um die Hyperparameter-Suche zu beschleunigen, und den gesamten Datensatz für das Modell-Training.*


#### Aufgabe 4.2 - Visualisierung der Modell-Vorhersage (10 Punkte)
Erstellen Sie ein `100 x 100` Grid aus Längen-/Breitengradwerten unter Verwendung der minimalen und maximalen Werte in Ihrem Datensatz. Visualisieren Sie die Vorhersagen des KNN-Modells **für drei verschiedene Parkgebühren** - 3, 5 und 7 Euro. Verwenden Sie dazu die Funktion `plot_map`. Färben Sie den Marker entsprechend der Modell-Vorhersage. Beschreiben Sie mindestens 2 visuelle Veränderungen des vorhergesagten Nutzungsmusters auf der Karte.